src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java

```diff
@@ -0,0 +1,84 @@
+package com.google.devtools.build.lib.skyframe;
+
+import com.google.common.base.Function;
+import com.google.common.base.Predicate;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.devtools.build.lib.cmdline.Label;
+import com.google.devtools.build.lib.events.Event;
+import com.google.devtools.build.lib.events.ExtendedEventHandler;
+import com.google.devtools.build.skyframe.CycleInfo;
+import com.google.devtools.build.skyframe.CyclesReporter;
+import com.google.devtools.build.skyframe.SkyKey;
+import java.util.Optional;
+
+public class RegisteredToolchainsCycleReporter implements CyclesReporter.SingleCycleReporter {
+
+  private static final Predicate<SkyKey> IS_REGISTERED_TOOLCHAINS_SKY_KEY =
+      SkyFunctions.isSkyFunction(SkyFunctions.REGISTERED_TOOLCHAINS);
+
+  private static final Predicate<SkyKey> IS_CONFIGURED_TARGET_SKY_KEY =
+      SkyFunctions.isSkyFunction(SkyFunctions.CONFIGURED_TARGET);
+
+  private static final Predicate<SkyKey> IS_TOOLCHAIN_RESOLUTION_SKY_KEY =
+      SkyFunctions.isSkyFunction(SkyFunctions.TOOLCHAIN_RESOLUTION);
+
+  @Override
+  public boolean maybeReportCycle(
+      SkyKey topLevelKey,
+      CycleInfo cycleInfo,
+      boolean alreadyReported,
+      ExtendedEventHandler eventHandler) {
+    ImmutableList<SkyKey> pathToCycle = cycleInfo.getPathToCycle();
+    ImmutableList<SkyKey> cycle = cycleInfo.getCycle();
+    if (pathToCycle.isEmpty()) {
+      return false;
```

Comment:
why is this? is it b/c we don't expect the SkyKeys of relevance to be top-level keys of a skyframe evaluation?

please add a comment.

Reply:
I copied it from the skylark cycle reporter, but actually this shouldn't care, so I've removed the check.

Reply:
yes, i noticed that. i'd have the same comment about the skylark cycle reporter code :p